### PR TITLE
Package states backend - part1 - mostly for non-builtin

### PR DESCRIPTION
**Pull request by pinzart90**
What this PR does:
Adds backend structures to support the new states.
Adds some of the UI elements needed (resource strings, updated context menu and the state labels)

What this PR does not do:
Does not implement the unload/reload process of built-in packages
Does not enable multiple packages with the same name to be shown (built in and non built in)

image

As a point of reference 
image
https://www.figma.com/file/ArPq0XKmHfowMV5BiL3G33/DynamoPreference-PackageManager?node-id=711%3A1291

### CODE REVIEW

**Comment by mjkkirschner** [comment]
this has a bunch of test failures (53)
https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/2582/

**Comment by pinzart90** [comment]
@Jingyi-Wen 
Can you confirm that the different length labels (Loaded, Error, Scheduled to be unloaded) look ok ?
image

**Comment by pinzart90** [comment]
@Jingyi-Wen 
Can you look at the following changes in the "More"(three dots) menu:
image

**Comment by pinzart90** [comment]
@Jingyi-Wen 
When trying to uninstall a package from the "Installed packages" window, a bunch of warning messages can pop up with references to "Uninstall" or "Installed". Do we need to change the wording on those ? or do they still work as is.
Example:
image

**Comment by Jingyi-Wen** [comment]
> @Jingyi-Wen
> Can you confirm that the different length labels (Loaded, Error, Scheduled to be unloaded) look ok ?
> If not, maybe we should look into badges ....or have the labels a fixed dimension...
> image

The labels look good to me!

**Comment by Jingyi-Wen** [comment]
> @Jingyi-Wen
> Can you look at the following changes in the "More"(three dots) menu ?
> Is the new wording better ? or is the old one still good?
> image

This is a great question. We should definitely try to keep a consistent wording to avoid confusion. Although I think the old wording does still make sense, since "uninstall" sounds more like user behavior while "loaded/unloaded" sounds more like Dynamo behavior. I'm leaning towards unifying all the wordings, but open to discussion if the team prefer the old one

**Comment by Jingyi-Wen** [comment]
> @Jingyi-Wen
> When trying to uninstall a package from the "Installed packages" window, a bunch of warning messages can pop up with references to "Uninstall" or "Installed". Do we need to change the wording on those ? or do they still work as is.
> Example:
> image

Same as above. Either way make sense to me, but a consistent wording might be better. Would be curious to see what others think:)

**Comment by mjkkirschner** [comment]
@pinzart90 seems like there are thousands of failures with this log:
``` 
MESSAGE:
SetUp : System.NullReferenceException : Object reference not set to an instance of an object.
TearDown : System.NullReferenceException : Object reference not set to an instance of an object.
+++++++++++++++++++
STACK TRACE:
   at Dynamo.Linting.LinterExtensionDescriptor.GetHashCode() in c:\WorkspaceDynamo\src\DynamoCore\Linting\LinterExtensionDescriptor.cs:line 56
   at System.Collections.Generic.ObjectEqualityComparer`1.GetHashCode(T obj)
   at System.Collections.Generic.HashSet`1.InternalGetHashCode(T item)
   at System.Collections.Generic.HashSet`1.AddIfNotPresent(T value)
   at Dynamo.Linting.LinterManager..ctor(IExtensionManager extensionManager) in c:\WorkspaceDynamo\src\DynamoCore\Linting\LinterManager.cs:line
   ```

maybe make sure you have latest master merged in, I'm not sure if that was an issue @saintentropy fixed in his latest PR or something on your branch.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageViewModel.cs [hunk 1]
```diff
@@ -148,6 +223,15 @@ private void Uninstall()
...
                 var pmExtension = dynModel.GetPackageManagerExtension();
                 Model.UninstallCore(dynModel.CustomNodeManager, pmExtension.PackageLoader, dynModel.PreferenceSettings);
+
+                if (DebugModes.IsEnabled("DynamoPackageStates"))
+                {
+                    RaisePropertyChanged(nameof(PackageLoadStateTooltip));
+                    RaisePropertyChanged(nameof(PackageLoadStateText));
+                } else
+                {
+                    RaisePropertyChanged("EnableOldMarkedForUnistallState");
```
Comment (pinzart90) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageViewModel.cs:233]: this fixes a regression since the introduction of the debug mode

File: src/DynamoCoreWpf/Controls/InstalledPackagesControl.xaml [hunk 2]
```diff
@@ -67,13 +67,24 @@
...
+                                                    <Border Background="DimGray"
+                                                        HorizontalAlignment="Stretch"
+                                                        CornerRadius="5"
+                                                        ToolTip="{Binding Path=PackageLoadStateTooltip}"
+                                                        Margin="5,0,0,0"
+                                                        Padding="5,0,5,0">
+                                                        <TextBlock Text="{Binding Path=PackageLoadStateText}"
+                                                                TextTrimming="CharacterEllipsis"
+                                                                FontSize="10"
+                                                                FontFamily="Artifakt Element"
```
Comment (pinzart90) [inline src/DynamoCoreWpf/Controls/InstalledPackagesControl.xaml:81]: font still does not look quite right

File: src/DynamoPackages/Package.cs [hunk 3]
```diff
@@ -31,6 +30,41 @@ public string Name
...
+        private StateTypes state = StateTypes.None;// Default to None type.
+
+        internal void SetAsLoaded() { state = StateTypes.Loaded; errorMessage = ""; }
+        internal void SetAsError(string msg = "") { state = StateTypes.Error; errorMessage = msg; }
+        internal void SetAsUnloaded() { state = StateTypes.Unloaded; errorMessage = ""; }
+
+        internal void SetScheduledForDeletion() { scheduledState = ScheduledTypes.ScheduledForDeletion; }
+        internal void SetScheduledForUnload() { scheduledState = ScheduledTypes.ScheduledForUnload; }
+        internal void SetScheduledForLoad() { scheduledState = ScheduledTypes.ScheduledForLoad; }
+        internal void ResetScheduledState() { scheduledState = ScheduledTypes.None; }
```
Comment (pinzart90) [inline src/DynamoPackages/Package.cs:56]: I went with the option of using explicit setter methods (ex. SetAsUnloaded) because it makes it easier to use the Error state + error message.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageViewModel.cs [hunk 4]
```diff
@@ -30,6 +31,80 @@ public bool HasAdditionalAssemblies
             get { return Model.LoadedAssemblies.Any(x => !x.IsNodeLibrary); }
         }
 
+        [Obsolete("This is not final. Please do not use it")]
```
Comment (pinzart90) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageViewModel.cs:34]: Will remove the obsolete attribute once this PR passes initial review

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageViewModel.cs [hunk 5]
```diff
@@ -30,6 +31,80 @@ public bool HasAdditionalAssemblies
...
+        public string PackageViewContextMenuUnmarkUninstallText
+        {
+            get { return Model.BuiltInPackage ? Resources.PackageContextMenuUnmarkUnloadPackageText : Resources.PackageContextMenuUnmarkDeletePackageText; }
+        }
+
+        [Obsolete("This is not final. Please do not use it")]
+        public string PackageViewContextMenuUnmarkUninstallTooltip
+        {
+            get { return Model.BuiltInPackage ? Resources.PackageContextMenuUnmarkUnloadPackageTooltip : Resources.PackageContextMenuUnmarkDeletePackageTooltip; }
+        }
```
Comment (pinzart90) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageViewModel.cs:106]: I did not use a converter (instead of these public properties) because it did not seem to simplify the situation very much

File: src/DynamoCoreWpf/Properties/Resources.en-US.resx [hunk 6]
```diff
@@ -2504,4 +2492,74 @@ Uninstall the following packages: {0}?</value>
...
+  <data name="PackageStateUnloadedTooltip" xml:space="preserve">
+    <value>Unloaded.
+This package has not been loaded because another conflicting package was loaded before it.</value>
+  </data>
+  <data name="PackageStateScheduledForUnloadTooltip" xml:space="preserve">
+    <value>Scheduled for Unload.
+This package will be unloaded after the next Dynamo restart.</value>
+  </data>
+  <data name="PackageStateScheduledForLoadTooltip" xml:space="preserve">
+    <value>Scheduled for Load.
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/Properties/Resources.en-US.resx:2534]: @jasonstratton what is this state for again? Can't packages be loaded immediately? Why do they need to be scheduled to load?

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageViewModel.cs [hunk 7]
```diff
@@ -30,6 +31,80 @@ public bool HasAdditionalAssemblies
...
+                        break;
+                }
+
+                switch (Model.LoadState.State)
+                {
+                    case PackageLoadState.StateTypes.Unloaded: return Resources.PackageStateUnloaded;
+                    case PackageLoadState.StateTypes.Loaded: return Resources.PackageStateLoaded;
+                    case PackageLoadState.StateTypes.Error: return Resources.PackageStateError;
+                    default:
+                        return "Unkonwn package state";
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageViewModel.cs:54]: typo: `unknown`

Comment (aparajit-pratap) [reply src/DynamoCoreWpf/ViewModels/PackageManager/PackageViewModel.cs:54]: Maybe this needs to be a resource string too.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageViewModel.cs [hunk 8]
```diff
@@ -30,6 +31,80 @@ public bool HasAdditionalAssemblies
...
+                        break;
+                }
+
+                switch (Model.LoadState.State)
+                {
+                    case PackageLoadState.StateTypes.Unloaded: return Resources.PackageStateUnloadedTooltip;
+                    case PackageLoadState.StateTypes.Loaded: return Resources.PackageStateLoadedTooltip;
+                    case PackageLoadState.StateTypes.Error: return string.Format(Resources.PackageStateErrorTooltip, Model.LoadState.ErrorMessage);
+                    default:
+                        return "Unkonwn package state";
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageViewModel.cs:79]: typo: unknown
Maybe this needs to be a resource string too.

File: src/DynamoCoreWpf/Properties/Resources.Designer.cs [hunk 9]
```diff
@@ -3925,6 +3889,78 @@ public class Resources {
...
+        ///   Looks up a localized string similar to Delete.
+        /// </summary>
+        public static string PackageContextMenuDeletePackageText {
+            get {
+                return ResourceManager.GetString("PackageContextMenuDeletePackageText", resourceCulture);
+            }
+        }
+        
+        /// <summary>
+        ///   Looks up a localized string similar to Deleting this package is only allowed if none of the types are currently in use.
```
[the remark this answers is not in the document]
Comment (pinzart90) [reply src/DynamoCoreWpf/Properties/Resources.Designer.cs:3902]: My impression was that this tooltip message referred to actually removing from disk (not scheduling or marking).
But we could update the message. Any ideas ? FYI @Jingyi-Wen 
Maybe - "Removes the package from Dynamo. Might require a restart for changes to take effect"..
Something similar for unloading ?

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageViewModel.cs (hunk 7 above)
Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/PackageManager/PackageViewModel.cs:54]: This is not supposed to be a valid state. It is only useful for debugging ... 
Do you think it is useful to be localized ?

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageViewModel.cs (hunk 8 above)
Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/PackageManager/PackageViewModel.cs:79]: This is not supposed to be a valid state. It is only useful for debugging ... 
Do you think it is useful to be localized ?

File: src/DynamoCoreWpf/Properties/Resources.en-US.resx (hunk 6 above)
Comment (pinzart90) [reply src/DynamoCoreWpf/Properties/Resources.en-US.resx:2534]: I guess the scenario is when Immediate Loading cannot happen because of conflict with an existing Package (that in turn cannot be deleted immediately)

File: src/DynamoCoreWpf/Properties/Resources.Designer.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Jingyi-Wen) [reply src/DynamoCoreWpf/Properties/Resources.Designer.cs:3902]: Changing the message makes sense. Can we possibly avoid using "might" though? It's important to indicate what would happen for certain

**Review by sm6srw — APPROVED** [review]
There are some open questions about some of the strings but after that is resolved: LGTM!

**Comment by aparajit-pratap** [comment]
Look like there are 2 test failures:
image

File: src/DynamoCoreWpf/Properties/Resources.Designer.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (pinzart90) [reply src/DynamoCoreWpf/Properties/Resources.Designer.cs:3902]: **Unload tooltip:**
Unloads this package so that it will not be visible in Dynamo anymore. The package will still exist in the Dynamo packages locations.

**Delete tooltip:**
Deletes this package from the Dynamo package locations. Once deleted, it will not be visible in Dynamo anymore.

[the remark this answers is not in the document]
Comment (pinzart90) [reply src/DynamoCoreWpf/Properties/Resources.Designer.cs:3902]: @Jingyi-Wen @aparajit-pratap would that work ?

**Comment by pinzart90** [comment]
> Look like there are 2 test failures:
> image

These pass locally. I think the build machine might be in a bad state...

File: src/DynamoCoreWpf/Properties/Resources.Designer.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Jingyi-Wen) [reply src/DynamoCoreWpf/Properties/Resources.Designer.cs:3902]: Make sense to me!

[the remark this answers is not in the document]
Comment (aparajit-pratap) [reply src/DynamoCoreWpf/Properties/Resources.Designer.cs:3902]: Looks fine to me. @Amoursol ?

[the remark this answers is not in the document]
Comment (Amoursol) [reply src/DynamoCoreWpf/Properties/Resources.Designer.cs:3902]: LGTM team! Thanks for the loop in @aparajit-pratap

[the remark this answers is not in the document]
Comment (mjkkirschner) [reply src/DynamoCoreWpf/Properties/Resources.Designer.cs]: @pinzart90 by "visible" do you mean visible in the node library? Or visible in the installed packages view? (my understanding is that unloaded pacakges will still show up in the packages view - but with an "unloaded" tag)

[the remark this answers is not in the document]
Comment (pinzart90) [reply src/DynamoCoreWpf/Properties/Resources.Designer.cs]: @mjkkirschner  I was thinking about the node library.
Should I just be specific about the node library ? That may be a bit complicated for regular users ... ?

[the remark this answers is not in the document]
Comment (pinzart90) [reply src/DynamoCoreWpf/Properties/Resources.Designer.cs]: Or should I just say:

**Unload tooltip:**
Unload this package so that it will not be loaded into Dynamo anymore. The package will still exist in the Dynamo packages locations.

**Delete tooltip:**
Delete this package from the Dynamo package locations. Once deleted, it will not be visible in Dynamo anymore.

[the remark this answers is not in the document]
Comment (mjkkirschner) [reply src/DynamoCoreWpf/Properties/Resources.Designer.cs]: I'm a little late to the conversation - so please disregard if you are happy with the current state.
but for your consideration:

**Unload tooltip:**
Unload this package so that it will not be loaded into Dynamo. The package will not be deleted from disk.

**Delete tooltip:**
Delete this package from disk. Once deleted, it will not be loaded into Dynamo anymore until it is re-downloaded.

File: src/DynamoCore/Configuration/PathManager.cs [hunk 10]
```diff
@@ -104,20 +104,20 @@ private IEnumerable<string> RootDirectories
...
         /// </summary>
-        internal string BuiltinPackagesDirectory
+        internal static string BuiltinPackagesDirectory
         {
             get
             {
                 if (builtinPackagesDirectory == null)
                 {
-                    builtinPackagesDirectory = Path.Combine(Path.GetDirectoryName(Assembly.GetAssembly(GetType()).Location), builtinPackagesDirName, @"Packages");
+                    builtinPackagesDirectory = Path.Combine(Path.GetDirectoryName(Assembly.GetAssembly(typeof(PathManager)).Location), builtinPackagesDirName, @"Packages");
```
Comment (mjkkirschner) [inline src/DynamoCore/Configuration/PathManager.cs:120]: curious - why was this necessary? Because of the change to static?

File: src/DynamoCoreWpf/Controls/InstalledPackagesControl.xaml (hunk 2 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/Controls/InstalledPackagesControl.xaml:81]: is Artifakt actually being used?

File: src/DynamoPackages/Package.cs [hunk 11]
```diff
@@ -31,6 +30,41 @@ public string Name
         }
     }
 
+    public class PackageLoadState
+    {
+        public enum StateTypes
```
Comment (mjkkirschner) [inline src/DynamoPackages/Package.cs:35]: I think if all this is going to be public it would be nice if this class, and all enums had /// summarys.

File: src/DynamoPackages/Package.cs [hunk 12]
```diff
@@ -31,6 +30,41 @@ public string Name
...
+        internal void SetAsError(string msg = "") { state = StateTypes.Error; errorMessage = msg; }
+        internal void SetAsUnloaded() { state = StateTypes.Unloaded; errorMessage = ""; }
+
+        internal void SetScheduledForDeletion() { scheduledState = ScheduledTypes.ScheduledForDeletion; }
+        internal void SetScheduledForUnload() { scheduledState = ScheduledTypes.ScheduledForUnload; }
+        internal void SetScheduledForLoad() { scheduledState = ScheduledTypes.ScheduledForLoad; }
+        internal void ResetScheduledState() { scheduledState = ScheduledTypes.None; }
+
+        // The current load state of the Package.
+        public StateTypes State { get { return state; } }
```
Comment (mjkkirschner) [inline src/DynamoPackages/Package.cs:59]: these properties too please- `///` summary

File: src/DynamoPackages/Package.cs [hunk 13]
```diff
@@ -107,71 +146,21 @@ public bool TypesVisibleInManager
...
-        public bool EnablePackageStates
-        {
-            get { return DebugModes.IsEnabled("DynamoPackageStates"); }
-        }
-
-        internal enum PackageStates
+        internal bool BuiltInPackage
         {
-            Loaded, Unloaded, PendingUnload, Error
+            get { return RootDirectory.Contains(PathManager.BuiltinPackagesDirectory); }
```
Comment (mjkkirschner) [inline src/DynamoPackages/Package.cs:151]: 👍

**Comment by mjkkirschner** [comment]
@jasonstratton @pinzart90 I think it's important to make sure the "what does this PR not do"
```
What this PR does not do:
Does not implement the unload/reload process of built-in packages
Does not enable multiple packages with the same name to be shown (built in and non built in)
```
section is covered by some existing task - I guess 
bullet one should be covered by: 
https://jira.autodesk.com/browse/DYN-3727
and bullet 2 might be covered by https://jira.autodesk.com/browse/DYN-3726 - though it's not really clear.

File: src/DynamoCoreWpf/Controls/InstalledPackagesControl.xaml (hunk 2 above)
Comment (pinzart90) [reply src/DynamoCoreWpf/Controls/InstalledPackagesControl.xaml:81]: does not look like it from the UI snapshot.
But no parts of the new UI seem to actually use Artifakt...

File: src/DynamoCore/Configuration/PathManager.cs (hunk 10 above)
Comment (pinzart90) [reply src/DynamoCore/Configuration/PathManager.cs:120]: Correct, because of move to static
